halo2_debug/src/lib.rs

```diff
@@ -0,0 +1,44 @@
+use rand_core::block::BlockRng;
+use rand_core::block::BlockRngCore;
+use rand_core::OsRng;
+use tiny_keccak::Hasher;
+
+// One number generator, that can be used as a deterministic Rng, outputing fixed values.
+pub struct OneNg {}
+
+impl BlockRngCore for OneNg {
+    type Item = u32;
+    type Results = [u32; 16];
+
+    fn generate(&mut self, results: &mut Self::Results) {
+        for elem in results.iter_mut() {
+            *elem = 1;
+        }
+    }
+}
+
+pub fn one_rng() -> BlockRng<OneNg> {
+    BlockRng::<OneNg>::new(OneNg {})
+}
+
+// Random number generator for testing
+
+pub fn test_rng() -> OsRng {
+    OsRng
+}
+
+fn keccak_hex<D: AsRef<[u8]>>(data: D) -> String {
+    let mut hash = [0u8; 32];
+    let mut hasher = tiny_keccak::Keccak::v256();
+    hasher.update(data.as_ref());
+    hasher.finalize(&mut hash);
+    hex::encode(hash)
+}
+
+// Check the a test proof against a known hash
+// Note that this function is only called in CI in "cargo test --all-fetaures"
+pub fn assert_test_proof<D: AsRef<[u8]>>(hex: &str, data: D) {
+    if cfg!(all(feature = "thread-safe-region", not(coverage))) {
```

Comment:
I have two questions about this:
1. The `halo2_debug` crate doesn't have any feature, which makes me think this is never activated
2. What's the rationale behind using `thread-safe-region` for this?  When this feature is not enabled in the frontend, a region is not thread safe, and thus parallel witness generation is not supported; but it this feature shouldn't affect the determinism of the proof generation.

Reply:
> Cause I'm getting different proofs if I activate `test_mycircuit_full_legacy` with `thread-safe-region`!

I see, we should investigate why this is happening.

Also please take a look at my point 1 :)